multik-jvm/src/main/kotlin/org/jetbrains/kotlinx/multik/jvm/JvmLinAlg.kt

```diff
@@ -20,29 +29,222 @@ public object JvmLinAlg : LinAlg {
+                }
+            }
+        }
+
+        return result.pow(1 / power.toDouble())
+    }
+
+    private fun <T : Number> normGeneral(
+        mat: ImmutableMemoryView<T>, matOffset: Int, matStrides: IntArray,
+        n: Int, m: Int, power: Int,
+        consistent: Boolean
+    ): Double {
+        var result = 0.0
+
+        val (matStride_0, matStride_1) = matStrides
+
+        if (consistent) {
+            result = mat.sumOf { abs(it.toDouble()).pow(power) }
+        } else {
+            for (i in 0 until n) {
+                val matInd = i * matStride_0 + matOffset
+                for (k in 0 until m) {
+                    val elementDoubleAbsValue = abs(mat[matInd + k * matStride_1].toDouble())
+                    result += (elementDoubleAbsValue).pow(power)
+                }
+            }
+        }
+
+        return result.pow(1 / power.toDouble())
+    }
+//----------------- end of cases for norm method ----------------------
+
+
+
+
+    //-------------------------------solve linear system-----------------------------------
+    /**
+     * solves a*x = b where a lower or upper triangle square matrix
+     */
+    public fun solveTriangle(a: D2Array<Double>, b: D2Array<Double>, isLowerTriangle: Boolean = true): D2Array<Double> {
+        require(a.shape[1] == b.shape[0]) { "invalid arguments, a.shape[1] = ${a.shape[1]} != b.shape[0]=${b.shape[0]}" }
+        require(a.shape[0] == a.shape[1]) { "a should be a square matrix, matrix with shape (${a.shape[0]}, ${a.shape[1]}) given" }
+        val x = b.deepCopy()
+        for (i in 0 until x.shape[0]) {
+            for (j in 0 until x.shape[1]) {
+                x[i, j] /= a[i, i]
+            }
+        }
+        if (isLowerTriangle) {
+            for (i in 0 until x.shape[0]) {
+                for (k in i + 1 until x.shape[0]) {
+                    for (j in 0 until x.shape[1]) {
+                        x[k, j] -= a[k, i] * x[i, j] / a[k, k]
+                    }
+                }
+            }
+        } else {
+            for (i in x.shape[0] - 1 downTo 0) {
+                for (k in i - 1 downTo 0) {
+                    for (j in 0 until x.shape[1]) {
+                        x[k, j] -= a[k, i] * x[i, j] / a[k, k]
+                    }
+                }
+            }
+
+        }
+        return x
+    }
+
+    fun solveDouble(a: D2Array<Double>, b: D2Array<Double>, singularityErrorLevel: Double = 1e-7): D2Array<Double> {
+        require(a.shape[0] == a.shape[1] && a.shape[1] == b.shape[0])
+        val (P, L, U) = PLUCompressed(a)
+        val _b = b.deepCopy()
+
+        for (i in P.indices) {
+            if(P[i] != 0) {
+                _b[i] = _b[i + P[i]].deepCopy().also { _b[i + P[i]] = _b[i].deepCopy() }
+            }
+        }
+        for (i in 0 until U.shape[0]) {
+            if (abs(U[i, i]) < singularityErrorLevel) {
+                throw ArithmeticException("matrix a is singular or almost singular")
```

Comment:
Beginning of the error is capitalized.
It is also probably better to use "nearly singular matrix".